src/tls_relation.py

```diff
@@ -53,16 +53,13 @@ def get_provider_cert_with_hostname(
         """
         if len(self.certificates.certificate_requests) == 0:
             return None
-        provider_certificate, _ = self.certificates.get_assigned_certificate(
-            certificate_request=self.certificates.certificate_requests[0]
-        )
-        if not provider_certificate:
-            return None
-        if not provider_certificate.certificate:
-            return None
-        if provider_certificate.certificate.common_name != hostname:
-            return None
-        return provider_certificate
+
+        provider_certificates, _ = self.certificates.get_assigned_certificates()
+        for provider_cert in provider_certificates:
```

Comment:
This block looks like a good candidate for a comprehension list

Reply:
I'm not sure how a list comprehension would work here, as we need to return the first matching cert in the middle of the loop, can you elaborate a bit further?

Reply:
Hmm, I'm not sure about this since this means we'll go through the full list of certificates before return, while currently we return as soon as we get a match. What do you think?